adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		return "", 0, 0, fmt.Errorf("getAdmFromHuaweiAdsContent failed: %s", err)
+	}
+	return adm, adWidth, adHeight, nil
+}
+
+// banner ad
+func (a *adapter) extractAdmBanner(adType int32, content *content, bidType openrtb_ext.BidType, imp openrtb2.Imp) (adm string,
+	adWidth int64, adHeight int64, err error) {
+	if adType != banner {
+		return "", 0, 0, fmt.Errorf("extractAdmBanner: huaweiads response is not a banner ad")
+	}
+	var creativeType = content.Creativetype
+	if content.Creativetype > 100 {
+		creativeType = creativeType - 100
+	}
+	if creativeType == text || creativeType == bigPicture || creativeType == bigPicture2 ||
+		creativeType == smallPicture || creativeType == threeSmallPicturesText ||
+		creativeType == iconText || creativeType == gif {
+		return a.extractAdmPicture(content)
+	} else if creativeType == videoText || creativeType == video || creativeType == videoWithPicturesText {
+		return a.extractAdmVideo(adType, content, bidType, imp)
+	} else {
+		return "", 0, 0, fmt.Errorf("no banner support creativetype")
+	}
+}
+
+// native ad
+func (a *adapter) extractAdmNative(adType int32, content *content, bidType openrtb_ext.BidType, openrtb2Imp openrtb2.Imp) (adm string,
+	adWidth int64, adHeight int64, err error) {
+	if adType != native {
+		return "", 0, 0, fmt.Errorf("extractAdmNative: response is not a native ad")
+	}
+	if openrtb2Imp.Native == nil {
+		return "", 0, 0, fmt.Errorf("extractAdmNative: imp.Native is nil")
+	}
+	if openrtb2Imp.Native.Request == "" {
+		return "", 0, 0, fmt.Errorf("extractAdmNative: imp.Native.Request is empty")
+	}
+
+	var nativePayload nativeRequests.Request
+	if err := json.Unmarshal(json.RawMessage(openrtb2Imp.Native.Request), &nativePayload); err != nil {
+		return "", 0, 0, err
+	}
+
+	var nativeResult nativeResponse.Response
+	var linkObject nativeResponse.Link
+	if content.MetaData.ClickUrl != "" {
+		linkObject.URL = content.MetaData.ClickUrl
+	} else if content.MetaData.Intent != "" {
+		linkObject.URL, _ = getDecodeValue(content.MetaData.Intent)
+	}
+
+	nativeResult.Assets = make([]nativeResponse.Asset, 0, len(nativePayload.Assets))
+	var imgIndex = 0
+	var iconIndex = 0
+	for _, asset := range nativePayload.Assets {
+		var responseAsset nativeResponse.Asset
+		if asset.Title != nil {
+			var titleObject nativeResponse.Title
+			titleObject.Text, _ = getDecodeValue(content.MetaData.Title)
+			titleObject.Len = int64(len(titleObject.Text))
+			responseAsset.Title = &titleObject
+		} else if asset.Video != nil {
+			var videoObject nativeResponse.Video
+			var err error
+			if videoObject.VASTTag, adWidth, adHeight, err = a.extractAdmVideo(adType, content, bidType, openrtb2Imp); err != nil {
+				return "", 0, 0, err
+			}
+			responseAsset.Video = &videoObject
+		} else if asset.Img != nil {
+			var imgObject nativeResponse.Image
+			imgObject.URL = ""
+			imgObject.Type = asset.Img.Type
+			if asset.Img.Type == native1.ImageAssetTypeIcon {
+				if len(content.MetaData.Icon) > iconIndex {
```

Comment:
Please add test coverage for this condition.

Reply:
add nativeThreeImageIncludeIcon.json